src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs

```diff
@@ -1142,6 +1188,71 @@ protected virtual void ReportUnassigned(Symbol symbol, SyntaxNode node, int slot
             _alreadyReported[slot] = true;
+
+            return;
+
+            bool needsImplicitFieldInitialization()
+                => _requireOutParamsAssigned
+                && CurrentSymbol is MethodSymbol { MethodKind: MethodKind.Constructor, ContainingType.TypeKind: TypeKind.Struct };
+
+            bool useWarningForStructThis(Symbol thisParameter, int thisSlot)
+            {
+                if (!needsImplicitFieldInitialization())
+                {
+                    return false;
+                }
+
+                Debug.Assert(_emptyStructTypeCache
+                    .GetStructInstanceFields(thisParameter.ContainingType)
+                    .Any(field => VariableSlot(field, thisSlot) is var fieldSlot && (fieldSlot == -1 || !State.IsAssigned(fieldSlot))));
+
+                foreach (var field in _emptyStructTypeCache.GetStructInstanceFields(thisParameter.ContainingType))
+                {
+                    if (_emptyStructTypeCache.IsEmptyStructType(field.Type))
+                        continue;
+
+                    if (field is TupleErrorFieldSymbol)
+                        continue;
+
+                    int slot = VariableSlot(field, thisSlot);
+                    if (slot == -1 || !State.IsAssigned(slot))
+                    {
+                        Debug.Assert(CurrentSymbol.EnclosingThisSymbol().Type.GetMembers(field.Name).Contains(field, ReferenceEqualityComparer.Instance));
+                        AddImplicitlyInitializedField(field);
+                    }
+                }
+
+                return compilation.IsFeatureEnabled(MessageID.IDS_FeatureAutoDefaultStructs);
+            }
+
+            bool useWarningForStructField(int fieldSlot)
+            {
+                if (!needsImplicitFieldInitialization())
+                {
+                    return false;
+                }
+
+                var thisSlot = GetOrCreateSlot(CurrentSymbol.EnclosingThisSymbol(), createIfMissing: false);
+                while (fieldSlot != -1 && variableBySlot[fieldSlot].ContainingSlot is int containingSlot && containingSlot != thisSlot)
+                {
+                    if (containingSlot == fieldSlot)
+                    {
+                        // the offending field access is not contained in 'this'.
+                        return false;
+                    }
+                    fieldSlot = containingSlot;
+                }
+
+                // the field was contained in 'this' by some level of nesting.
+                // we now have a slot for the field directly contained in 'this'.
+                // should we handle nested fields here? https://github.com/dotnet/roslyn/issues/59890
+                if (fieldSlot != -1)
+                {
+                    AddImplicitlyInitializedField((FieldSymbol)variableBySlot[fieldSlot].Symbol);
+                }
+
+                return compilation.IsFeatureEnabled(MessageID.IDS_FeatureAutoDefaultStructs);
```

Comment:
Should we only do this feature check if `AddImplicitlyInitializedField` was called?